### PR TITLE
chore: utils errors refactor

### PR DESCRIPTION
**Pull request by bhflm**
***What***
Current is a stab to refactor existing `utils/errors.ts`, 

***Why***

In an effort to make the sdk more readable, this provides an enhancement for legibility of the current errors file, without the need of hardcoding existing/new errors and make it more _extendable_ for future releases. 

***How***

* Separate onto a constants.ts to categorize different kind of errors types. 
* Create util functions to match error strings and avoid repetition of several if statements with string matching
* Create custom error classes to manipulate errors differently, and separate them between Railgun Contract errors and General custom errors. 
* Adds tests for edge cases such as undefined, and ascii characters.

### CODE REVIEW

File: src/utils/errors/constants.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import { CustomErrorMapping } from "./types";
+
+export const STRING_PREFIX_AFTER_UNICODE_REPLACEMENT = 'y %';
+
+// eslint-disable-next-line no-useless-escape
```
Comment (mattgle) [inline src/utils/errors/constants.ts:5]: is this really needed?

**Review by mattgle — COMMENTED** [review]
Just a few comments, not really a MUST, but improvements IMO

File: src/utils/errors/error.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      return errorDef;
+    }
+  }
+  return null;
+}
+
+const isRailgunError = (cause: Error): boolean => cause.message.toLowerCase().includes('railgunsmartwallet')
+
+export const sanitizeError = (cause: Error): Error => {
+  if (isDefined(cause) && cause.message) {
```
Comment (emilbayes) [inline src/utils/errors/error.ts:26]: Personally, since the "else" here is shorter, early return would make the logic easier to read, eg.

```ts
if (cause?.message == null) return new Error('Unknown error. Please try again.', { cause });

if (isRailgunError(cause)) { /* ... */ }
```

Another question is, do we even need to check if `cause` is defined and has a `message` property? The type here says it must be an `Error`, so we perhaps only need to check the `message` being defined

File: src/utils/errors/constants.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+export const INVALID_ASCII_REGEX = /[^A-Za-z 0-9 \.,\?""!@#\$%\^&\*\(\)-_=\+;:<>\/\\\|\}\{\[\]`~]*/g;
```
Comment (emilbayes) [inline src/utils/errors/constants.ts:6]: A neat little trick (that begs a comment), but all the printable ascii characters are in the range from ` ` (space) to `~` (tilde), so the regex can be compressed to: `/[^ -~]+/g`. Is this better? I think so (with a comment explaining what it is) since it ensures no printable ascii chars were missed, despite being more obscure knowledge, where a list like the above requires you to check that all symbols were included.

File: src/utils/errors/constants.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { CustomErrorMapping } from "./types";
+
+export const STRING_PREFIX_AFTER_UNICODE_REPLACEMENT = 'y %';
```
Comment (emilbayes) [inline src/utils/errors/constants.ts:3]: Do we know what this means? Looking at the usage, I don't understand how the name of the constant here makes sense (I know you just migrated it from what it historically was)

File: src/utils/errors/error.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import { CustomErrorMapping, ErrorDefinition } from './types';
+import { STRING_PREFIX_AFTER_UNICODE_REPLACEMENT, RAILGUN_ERRORS, CUSTOM_ERRORS, INVALID_ASCII_REGEX } from './constants';
+import { isDefined } from '../util';
+
+const validAscii = (str: string) => {
```
Comment (emilbayes) [inline src/utils/errors/error.ts:5]: Now that we are refactoring, perhaps we should give this a more fitting name, perhaps `sanitizeAscii`

File: src/utils/errors/error.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    // If no error is matched we return the original sanitized error
+    const errorMessage = validAscii(cause.message).replace(
+      `:${STRING_PREFIX_AFTER_UNICODE_REPLACEMENT}`,
+      ': ',
+    );
+
+    return new Error(
+      errorMessage,
+      cause
```
Comment (emilbayes) [inline src/utils/errors/error.ts:49]: All these `cause` should be in an object since the usage was a shorthand for the `Error` `options.cause`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#cause

File: src/utils/index.ts [hunk 7]
```diff
@@ -2,7 +2,7 @@ export * from './artifact-v2';
 export * from './available-rpc';
 export * from './compare';
 export * from './fallback-provider';
-export * from './error';
+export * from './errors/error';
```
Comment (emilbayes) [inline src/utils/index.ts:5]: Perhaps this is an "internal module" so changing the path is insignificant, but if something is importing the error formatter directly, this is a breaking change

File: src/utils/errors/constants.ts (hunk 4 above)
Comment (bhflm) [reply src/utils/errors/constants.ts:3]: given the case that we keep it, a different name would be better IMHO,  `SANITIZED_MESSAGE_SEPARATOR` or similar

but now that I've mentioned, I think we could remove this and be shortened to something that looks like:  

```
const errorMessage = sanitizeAscii(cause.message);
```

since we're already properly sanitizing ASCII, and we don't need *that much* string manipulation besides this right??

File: src/utils/errors/error.ts (hunk 6 above)
Comment (bhflm) [reply src/utils/errors/error.ts:49]: good catch

File: src/utils/errors/constants.ts (hunk 4 above)
Comment (emilbayes) [reply src/utils/errors/constants.ts:3]: Yeah reading again I think either this string used to be placed somehow in the string or it's an artefact from something that emitted errors with that string, and someone replaced it. Anyhow, the purpose is unclear to me. Maybe @zy0n or @mattgle remember seeing this string in errors at some point?

Comment (mattgle) [reply src/utils/errors/constants.ts:3]: I have no idea tbh, but @mesquka might know

File: src/utils/errors/constants.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+// Matches any characters that are NOT in the printable ASCII range (space to tilde)
+// Printable ASCII characters are in the range of 32 (space) to 126 (tilde)
+export const INVALID_ASCII_REGEX = /[^ -~]+/g;
```
[the remark this answers is not in the document]
Comment (bhflm) [reply src/utils/errors/constants.ts:7]: lets keep it as it was before then 👍🏼

**Comment by mattgle** [comment]
What should we do with this?
CC: @emilbayes @bhflm
